From d4c85d96e50e0d01c89c6cf024089164e2031315 Mon Sep 17 00:00:00 2001
From: DarkLord1731 <darklord1731@gmail.com>
Date: Wed, 27 Jun 2018 22:33:05 +0530
Subject: [PATCH 1/3] audio: Add flag to opt in/out amplifier support

* We don't need this on devices without an amplifier
* This also silences logs like 'audio_hw_primary: Amplifier initialization failed' on devices without an amplifier

Change-Id: I18690d9428b52f9915e8cd6c2639d40dcf8699d9
---
 audio/Android.mk |  4 ++++
 audio/audio_hw.c | 33 +++++++++++++++++++++++++++++++++
 audio/audio_hw.h |  6 ++++++
 3 files changed, 43 insertions(+)

diff --git a/audio/Android.mk b/audio/Android.mk
index f369f7c..a932db4 100644
--- a/audio/Android.mk
+++ b/audio/Android.mk
@@ -53,6 +53,10 @@ LOCAL_C_INCLUDES += \
 LOCAL_CFLAGS := -Werror -Wall
 #LOCAL_CFLAGS += -DPREPROCESSING_ENABLED
 
+ifeq ($(DEVICE_HAS_AMPLIFIER_SUPPORT),true)
+LOCAL_CFLAGS += -DDEVICE_HAS_AMPLIFIER_SUPPORT
+endif
+
 LOCAL_MODULE := audio.primary.$(TARGET_BOOTLOADER_BOARD_NAME)
 
 LOCAL_MODULE_RELATIVE_PATH := hw
diff --git a/audio/audio_hw.c b/audio/audio_hw.c
index f97758c..24fbb72 100644
--- a/audio/audio_hw.c
+++ b/audio/audio_hw.c
@@ -202,6 +202,7 @@ static const struct string_to_enum out_channels_name_to_enum_table[] = {
 
 static struct audio_device *adev = NULL;
 
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
 static amplifier_device_t * get_amplifier_device(void)
 {
     if (adev)
@@ -333,6 +334,7 @@ static int amplifier_close(void)
 
     return 0;
 }
+#endif
 
 struct timespec time_spec_diff(struct timespec time1, struct timespec time0) {
     struct timespec ret;
@@ -927,7 +929,9 @@ static int enable_snd_device(struct audio_device *adev,
         }
 #endif /* DSP_POWEROFF_DELAY */
 
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
         amplifier_enable_devices(snd_device, true);
+#endif
 
         audio_route_apply_and_update_path(mixer_card->audio_route, snd_device_name);
     }
@@ -976,7 +980,10 @@ int disable_snd_device(struct audio_device *adev,
                 audio_route_apply_and_update_path(mixer_card->audio_route, out_snd_device_name);
             }
 
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
             amplifier_enable_devices(snd_device, false);
+#endif
+
 #ifdef DSP_POWEROFF_DELAY
             clock_gettime(CLOCK_MONOTONIC, &(mixer_card->dsp_poweroff_time));
 #endif /* DSP_POWEROFF_DELAY */
@@ -1094,9 +1101,11 @@ static int select_devices(struct audio_device *adev,
     usecase->in_snd_device = in_snd_device;
     usecase->out_snd_device = out_snd_device;
 
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
     /* Rely on amplifier_set_devices to distinguish between in/out devices */
     amplifier_set_input_devices(in_snd_device);
     amplifier_set_output_devices(out_snd_device);
+#endif
 
     return 0;
 }
@@ -2642,7 +2651,9 @@ static int out_standby(struct audio_stream *stream)
     lock_output_stream(out);
     if (!out->standby) {
         pthread_mutex_lock(&adev->lock);
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
         amplifier_output_stream_standby((struct audio_stream_out *) stream);
+#endif
         do_out_standby_l(out);
         pthread_mutex_unlock(&adev->lock);
     }
@@ -2776,7 +2787,9 @@ static int out_set_parameters(struct audio_stream *stream, const char *kvpairs)
         pthread_mutex_unlock(&adev->lock_inputs);
     }
 
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
     amplifier_set_parameters(parms);
+#endif
 
     if (out->usecase == USECASE_AUDIO_PLAYBACK_OFFLOAD) {
         parse_compress_metadata(out, parms);
@@ -2925,9 +2938,12 @@ static ssize_t out_write(struct audio_stream_out *stream, const void *buffer,
 #endif
         pthread_mutex_lock(&adev->lock);
         ret = start_output_stream(out);
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
         if (ret == 0) {
             amplifier_output_stream_start(stream, out->usecase == USECASE_AUDIO_PLAYBACK_OFFLOAD);
         }
+#endif
 
         /* ToDo: If use case is compress offload should return 0 */
         if (ret != 0) {
@@ -3331,7 +3347,11 @@ static int in_standby_l(struct stream_in *in)
     lock_input_stream(in);
     if (!in->standby) {
         pthread_mutex_lock(&adev->lock);
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
         amplifier_input_stream_standby((struct audio_stream_in *) in);
+#endif
+
         status = do_in_standby_l(in);
         pthread_mutex_unlock(&adev->lock);
     }
@@ -3482,9 +3502,13 @@ static ssize_t in_read(struct audio_stream_in *stream, void *buffer,
         }
         pthread_mutex_lock(&adev->lock);
         ret = start_input_stream(in);
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
         if (ret == 0) {
             amplifier_input_stream_start(stream);
         }
+#endif
+
         pthread_mutex_unlock(&adev->lock);
         pthread_mutex_unlock(&adev->lock_inputs);
 
@@ -4050,9 +4074,12 @@ static int adev_set_mode(struct audio_hw_device *dev, audio_mode_t mode)
     pthread_mutex_lock(&adev->lock);
     if (adev->mode != mode) {
         ALOGI("%s mode = %d", __func__, mode);
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
         if (amplifier_set_mode(mode) != 0) {
             ALOGE("Failed setting amplifier mode");
         }
+#endif
         adev->mode = mode;
 
         if ((mode == AUDIO_MODE_NORMAL) && adev->voice.in_call) {
@@ -4252,11 +4279,15 @@ static int adev_close(hw_device_t *device)
     struct audio_device *adev = (struct audio_device *)device;
     voice_session_deinit(adev->voice.session);
     audio_device_ref_count--;
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
     if (audio_device_ref_count == 0) {
         if (amplifier_close() != 0) {
             ALOGE("Amplifier close failed");
         }
     }
+#endif
+
     free(adev->snd_dev_ref_cnt);
     free_mixer_list(adev);
     free(device);
@@ -4375,9 +4406,11 @@ static int adev_open(const hw_module_t *module, const char *name,
         return -EINVAL;
     }
 
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
     if (amplifier_open() != -ENOENT) {
         ALOGE("Amplifier initialization failed");
     }
+#endif
 
     *device = &adev->device.common;
 
diff --git a/audio/audio_hw.h b/audio/audio_hw.h
index 4677f29..6b5d29f 100644
--- a/audio/audio_hw.h
+++ b/audio/audio_hw.h
@@ -21,7 +21,10 @@
 
 #include <cutils/list.h>
 #include <hardware/audio.h>
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
 #include <hardware/audio_amplifier.h>
+#endif
 
 #include <tinyalsa/asoundlib.h>
 #include <tinycompress/tinycompress.h>
@@ -408,7 +411,10 @@ struct audio_device {
 #endif
 
     pthread_mutex_t         lock_inputs; /* see note below on mutex acquisition order */
+
+#ifdef DEVICE_HAS_AMPLIFIER_SUPPORT
     amplifier_device_t      *amp;
+#endif
 };
 
 /*
-- 
2.17.1

